File: scripts/build_standalone.py
#!/usr/bin/env python3
"""Build one HTML file containing CSS, JavaScript and the complete dataset."""
from __future__ import annotations

import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DATA_DIR = ROOT / "data"
OUTPUT = ROOT / "valkompass-2026-standalone.html"
VERSIONED_OUTPUT = ROOT / "valkompass-2026-standalone-v0.4.html"


def read_json(name: str):
    return json.loads((DATA_DIR / name).read_text(encoding="utf-8"))


def safe_script_json(value) -> str:
    # Prevent a data string from ending the surrounding script element.
    return json.dumps(value, ensure_ascii=False, separators=(",", ":")).replace("</", "<\\/")


def safe_style_text(value: str) -> str:
    # Prevent an accidental literal closing style tag in embedded CSS.
    return value.replace("</style", "<\\/style")


def safe_script_text(value: str) -> str:
    # Prevent an accidental literal closing script tag in embedded JavaScript.
    return value.replace("</script", "<\\/script")


def main() -> None:
    html = (ROOT / "index.html").read_text(encoding="utf-8")
    css = (ROOT / "styles.css").read_text(encoding="utf-8")
    scoring = (ROOT / "scoring.js").read_text(encoding="utf-8")
    app = (ROOT / "app.js").read_text(encoding="utf-8")

    dataset = {
        "questions": read_json("questions.json"),
        "parties": read_json("parties.json"),
        "positions": read_json("positions.json"),
        "meta": read_json("meta.json"),
        "sources": read_json("source-register.json"),
    }

    # The standalone build puts both modules in one module scope.
    scoring = re.sub(r"^export\s+", "", scoring, flags=re.MULTILINE)
    app = re.sub(
        r'^import\s*\{[\s\S]*?\}\s*from\s*["\']\.\/scoring\.js["\'];\s*',
        "",
        app,
        count=1,
    )

    html, stylesheet_replacements = re.subn(
        r'<link\s+rel=["\']stylesheet["\']\s+href=["\']styles\.css["\']\s*/?>',
        lambda _match: f"<style>\n{safe_style_text(css)}\n</style>",
        html,
        count=1,
        flags=re.IGNORECASE,
    )
    if stylesheet_replacements != 1:
        raise RuntimeError("Kunde inte bädda in styles.css: stylesheet-taggen hittades inte exakt en gång.")
    embedded = f"<script>window.__VALKOMPASS_DATA__={safe_script_json(dataset)};</script>"
    application_source = safe_script_text(f"{scoring}\n\n{app}")
    application = (
        "<script>\n"
        "(async () => {\n"
        f"{application_source}\n"
        "})().catch((error) => console.error(error));\n"
        "</script>"
    )
    html, script_replacements = re.subn(
        r'<script\s+type=["\']module["\']\s+src=["\']app\.js["\']\s*></script>',
        lambda _match: f"{embedded}\n  {application}",
        html,
        count=1,
        flags=re.IGNORECASE,
    )
    if script_replacements != 1:
        raise RuntimeError("Kunde inte bädda in app.js: module-scriptet hittades inte exakt en gång.")
    html = re.sub(
        r"(<title>\s*Öppen\s+Valkompass\s+2026)(\s*</title>)",
        r"\1 – fristående\2",
        html,
        count=1,
        flags=re.IGNORECASE,
    )

    if 'href="styles.css"' in html or 'src="app.js"' in html:
        raise RuntimeError("Standalone-bygget innehåller fortfarande externa lokala beroenden.")

    OUTPUT.write_text(html, encoding="utf-8")
    VERSIONED_OUTPUT.write_text(html, encoding="utf-8")
    print(f"Built {OUTPUT} ({OUTPUT.stat().st_size:,} bytes)")
    print(f"Built {VERSIONED_OUTPUT} ({VERSIONED_OUTPUT.stat().st_size:,} bytes)")


if __name__ == "__main__":
    main()

File: scripts/build-standalone.mjs
#!/usr/bin/env node
import { readFile, writeFile } from "node:fs/promises";
import { resolve } from "node:path";
import process from "node:process";

const root = resolve(new URL("..", import.meta.url).pathname);
const output = resolve(root, "valkompass-2026-standalone.html");

const readText = (name) => readFile(resolve(root, name), "utf8");
const readJson = async (name) => JSON.parse(await readText(name));
const escapeScriptEnd = (text) => text.replaceAll("</script", "<\\/script");
const escapeStyleEnd = (text) => text.replaceAll("</style", "<\\/style");

const [html, css, scoringSource, appSource, questions, parties, positions, meta, sources] = await Promise.all([
  readText("index.html"),
  readText("styles.css"),
  readText("scoring.js"),
  readText("app.js"),
  readJson("data/questions.json"),
  readJson("data/parties.json"),
  readJson("data/positions.json"),
  readJson("data/meta.json"),
  readJson("data/source-register.json"),
]);

const scoring = scoringSource.replaceAll(/\bexport\s+(?=function|const|class)/g, "");
const app = appSource.replace(
  /^import\s*\{[\s\S]*?\}\s*from\s*["']\.\/scoring\.js["'];\s*/,
  "",
);
if (app === appSource) {
  throw new Error("Kunde inte ta bort scoring-importen ur app.js. Bygget avbröts.");
}

const embedded = JSON.stringify({ questions, parties, positions, meta, sources })
  .replaceAll("<", "\\u003c")
  .replaceAll("\u2028", "\\u2028")
  .replaceAll("\u2029", "\\u2029");

let standalone = html.replace(
  '<link rel="stylesheet" href="styles.css">',
  `<style>\n${escapeStyleEnd(css)}\n</style>`,
);
standalone = standalone.replace(
  '<script type="module" src="app.js"></script>',
  `<script>window.__VALKOMPASS_DATA__=${embedded};</script>\n<script>\n(async () => {\n${escapeScriptEnd(scoring)}\n\n${escapeScriptEnd(app)}\n})().catch((error) => console.error(error));\n</script>`,
);
standalone = standalone.replace(
  "En öppen, källgranskad och reproducerbar svensk valkompass inför riksdagsvalet 2026.",
  "En fristående, öppen och reproducerbar svensk valkompass inför riksdagsvalet 2026.",
);
standalone = `<!-- Genererad fil. Ändra källfilerna och kör npm run build. Dataset ${meta.datasetVersion}, SHA-256 ${meta.dataFingerprintSha256}. -->\n${standalone}`;

await writeFile(output, standalone, "utf8");
console.log(`Byggde ${output}`);
console.log(`Dataset ${meta.datasetVersion} · ${questions.length} frågor · ${positions.length} positionsrader`);
